Lib/test/test_capi.py

```diff
@@ -333,6 +333,39 @@ def test_negative_refcount(self):
                          br'_Py_NegativeRefcount: Assertion failed: '
                          br'object has negative ref count')
 
+    def test_trashcan_subclass(self):
+        # bpo-35983: Check that the trashcan mechanism for "list" is NOT
+        # activated when its tp_dealloc is being called by a subclass
+        from _testcapi import MyList
+        L = None
+        for i in range(1000):
+            L = MyList((L,))
+
+    def test_trashcan_python_class(self):
+        # Check that the trashcan mechanism works properly for a Python
+        # subclass of a class using the trashcan (list in this test)
+        class PyList(list):
+            # Count the number of PyList instances to verify that there is
+            # no memory leak
+            num = 0
+            def __init__(self, *args):
+                __class__.num += 1
+                super().__init__(*args)
+            def __del__(self):
+                __class__.num -= 1
+
+        for parity in (0, 1):
+            L = None
+            for i in range(2**20):
```

Comment:
Is `2**20` really required to exercise this?

Reply:
> Is `2**20` really required to exercise this?

Yes: a typical stack on Linux is 8MB, so you need to use `2**23 / S` iterations, where `S` is the size of a stack frame. As bare minimum, the stack frame needs to contain the return address, so that's 8 bytes on a 64-bit system. I think it's possible to actually achieve this minimum, when omitting the frame pointer and storing all variables in registers.